presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java

```diff
@@ -368,66 +369,83 @@ protected Node visitShowPartitions(ShowPartitions showPartitions, Void context)
-
-                SELECT
-                  partition_number
-                , max(CASE WHEN partition_key = 'ds' THEN partition_value END) ds
-                , max(CASE WHEN partition_key = 'cluster_name' THEN partition_value END) cluster_name
-                FROM ...
-                GROUP BY partition_number
-
-                The values are also cast to the type of the partition column.
-                The query is then wrapped to allow custom filtering and ordering.
-            */
-
-            ImmutableList.Builder<SelectItem> selectList = ImmutableList.builder();
-            ImmutableList.Builder<SelectItem> wrappedList = ImmutableList.builder();
-            selectList.add(unaliasedName("partition_number"));
-            for (ColumnHandle columnHandle : partitionColumns) {
-                ColumnMetadata column = metadata.getColumnMetadata(session, tableHandle, columnHandle);
-                Expression key = equal(identifier("partition_key"), new StringLiteral(column.getName()));
-                Expression value = caseWhen(key, identifier("partition_value"));
-                value = new Cast(value, column.getType().getTypeSignature().toString());
-                Expression function = functionCall("max", value);
-                selectList.add(new SingleColumn(function, new Identifier(column.getName())));
-                wrappedList.add(unaliasedName(column.getName()));
+
+            DiscretePredicates discretePredicates = layout.get().getDiscretePredicates()
+                    .orElseThrow(() -> new SemanticException(NOT_SUPPORTED, showPartitions, "Table does not have partition columns: %s", table));
+
+            List<TupleDomain<ColumnHandle>> predicates = ImmutableList.copyOf(discretePredicates.getPredicates());
+            if (predicates.isEmpty()) {
+                return simpleQuery(
+                        selectList(identifier("no partitions")),
+                        aliased(
+                                values(row(new NullLiteral())),
+                                "x",
+                                ImmutableList.of("no partitions")));
             }
 
-            Query query = simpleQuery(
-                    selectAll(selectList.build()),
-                    from(table.getCatalogName(), TABLE_INTERNAL_PARTITIONS),
-                    Optional.of(logicalAnd(
-                            equal(identifier("table_schema"), new StringLiteral(table.getSchemaName())),
-                            equal(identifier("table_name"), new StringLiteral(table.getObjectName())))),
-                    Optional.of(new GroupBy(false, ImmutableList.of(new SimpleGroupBy(ImmutableList.of(identifier("partition_number")))))),
-                    Optional.empty(),
-                    Optional.empty(),
-                    Optional.empty());
+            Map<ColumnHandle, String> columnHandleToName = ImmutableBiMap.copyOf(metadata.getColumnHandles(session, tableHandle)).inverse();
```

Comment:
No need to extract column handles from predicate if you use: `ConnectorTablePartitioning#partitioningColumns` (and if it's suitable).

Reply:
Didn't work